codeU2.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+class Tree():
+    
+    """Defines the class for the tree:
+    self.children contains a dict of pairs (node:(left_child,right_child))
+    self.head contains the value of the tree head
+    None denotes the absence if the node
+    """
+    
+    def __init__(self,children,head):
+        self.children = children;
+        self.head = head;
+
+def dfs_path(tree_, start, end, path=None):
+
+    """Yields all full paths from start node
+    to end node, excluding the end node
+    """
+    
+    tree = tree_.children
+    stack = [(start, [start])]
+    while stack:
+        (vertex, path) = stack.pop()
+        if vertex is None:
+            continue
+        for next in set(tree[vertex]) - set(path):
+            if next == end:
+                yield path
+            else:
+                stack.append((next, path+[next]))
+
+def get_ancestors(tree_, node):
+
+    """Returns the ancestors from head to a given node
+    """
+
+    head = tree_.head
+    if not head:
```

Comment:
Just food for thought - does `dfs_path` handle `head=None` and `head=node` cases? Should it? If these checks are required (functionally) at all, then which function should they belong to?

It's a different story if you want to log these conditions for visibility into how your algorithm works for debugging. In that case of course you can log interesting conditions wherever you need to know of them :)